src/mlpack/methods/ann/vanilla_rbm_impl.hpp

```diff
@@ -0,0 +1,232 @@
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+
+// Intialise a linear layer and sigmoid layer
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::RBM(
+    arma::mat predictors,
+    InitializationRuleType initializeRule,
+    VisibleLayerType visible,
+    HiddenLayerType hidden,
+    const size_t numSteps,
+    const bool persistence):
+    visible(visible),
+    hidden(hidden),
+    initializeRule(initializeRule),
+    numSteps(numSteps),
+    persistence(persistence),
+    reset(false)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::Reset()
+{
+  // Since we are dealing with a static stucture no need for visitor
+  // Reset the parameters of all the layers
+  size_t weight = 0;
+  weight+= visible.Parameters().n_elem;
+  parameter.set_size(weight, 1);
+  parameter.zeros();
+  initializeRule.Initialize(parameter, parameter.n_elem, 1);
+  visible.Parameters() = arma::mat(parameter.memptr(), weight, 1, false, false);
+  hidden.Parameters() = arma::mat(parameter.memptr(), weight, 1, false, false);
+
+  visible.Reset();
+  hidden.Reset();
+  reset = true;
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+template<template<typename> class Optimizer>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Train(const arma::mat& predictors, Optimizer<NetworkType>& optimizer)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+
+  if (!reset)
+  {
+    Reset();
+  }
+
+  // Train the model.
+  Timer::Start("rbm_optimization");
+  optimizer.Optimize(parameter);
+  Timer::Stop("rbm_optimization");
+}
+
+ /* 
+  * This function calculates
+  * the free energy of the model
+  */
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+double RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    FreeEnergy(arma::mat input)
+{
+  arma::mat output;
+  visible.ForwardPreActivation(std::move(input), std::move(output));
+  SoftplusFunction::Fn(output, output);
+  return  -(arma::accu(output) + arma::dot(input, visible.Bias()));
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+double RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Evaluate(const arma::mat& /* parameters*/, const size_t i)
+{
+  double freeEnergy, freeEnergyChain;
+  arma::mat negativeSample;
+  arma::vec currentInput = predictors.unsafe_col(i);
+  // Do not use unsafe col predictor as we change the input
+  freeEnergy = FreeEnergy(currentInput);
```

Comment:
Moreover if you apply `FreeEnergy` to `arma::vec` you should add a template parameter in order to prevent copying.

For example:
```
template<typename VectorType>
FreeEnergy(VectorType& input)
```